syne_tune/results_callback.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+from typing import Dict, Any, Optional
+from time import perf_counter
+import copy
+import pandas as pd
+
+from syne_tune.backend.trial_status import Trial
+from syne_tune.constants import (
+    ST_DECISION,
+    ST_TRIAL_ID,
+    ST_STATUS,
+    ST_TUNER_TIME,
+    ST_RESULTS_DATAFRAME_FILENAME,
+    ST_FINAL_RESULTS_FILENAME,
+)
+from syne_tune.tuner_callback import TunerCallback
+from syne_tune.util import RegularCallback, dump_json_with_numpy
+
+
+class FinalResultsComposer:
+    """
+    Base class for ``final_results_composer`` argument in
+    :class:`StoreResultsCallback`. Extracts final results at the end of
+    tuning and returns them as a dictionary which can be JSON-serialized.
+    This is for final results, which are not time-stamped.
+
+    Why don't we use a lambda function instead? We would like the tuner,
+    with all its dependent objects, to be dill serializable, and lambda
+    functions are not.
+    """
+
+    def __call__(self, tuner) -> Optional[Dict[str, Any]]:
+        """
+        Called in :meth:`StoreResultsCallback.on_tuning_end`. The dictionary
+        returned is stored as JSON file.
+
+        :param tuner: :class:`~syne_tune.Tuner` object from which final results
+            are extracted
+        :return: Dictionary (JSON-serializable) to be written, or ``None`` if
+            nothing should be written
+        """
+        raise NotImplementedError
+
+
+class StoreResultsCallback(TunerCallback):
+    """
+    Default implementation of :class:`~TunerCallback` which records all
+    reported results, and allows to store them as CSV file.
+
+    :param add_wallclock_time: If ``True``, wallclock time since call of
+        ``on_tuning_start`` is stored as
+        :const:`~syne_tune.constants.ST_TUNER_TIME`.
+    :param final_results_composer: Optional. If given, this is called in
+        :meth:`on_tuning_end`, and the resulting dictionary is written as
+        ``{tuner.tuner_path}/{ST_FINAL_RESULTS_FILENAME}``.
+    :param store_final_results_intermediately: If ``True`` and
```

Comment:
Do you mean `intermediately` or perhaps `intermittently` or `immediately`?

As a customer, I might be confused, because are they really "final" results if the tuning is still in-progress? Are they not "intermediate" results in some sense?

Reply:
I'll change it to "periodically". Note we also store `results.csv.zip` periodically, so if an experiment crashes after 10 hours, you still have something.